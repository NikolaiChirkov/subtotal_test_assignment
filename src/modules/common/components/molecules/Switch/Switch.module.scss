.root {
  display: flex;
  flex-flow: row nowrap;
  border: 1px solid white;
  border-radius: 16px;

  &__option {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
    height: 100%;
    padding: 8px 16px;

    &:first-child {
      border-top-left-radius: 16px;
      border-bottom-left-radius: 16px;
    }

    &:last-child {
      border-top-right-radius: 16px;
      border-bottom-right-radius: 16px;
    }
  }
}

.current {
  background-color: rgb(29, 29, 29);
}
