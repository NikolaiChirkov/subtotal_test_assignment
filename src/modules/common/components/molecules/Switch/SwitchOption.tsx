import React from "react";

export type SwitchOptionType = {
  title: string;
  className?: string;
  onClick: () => void;
};

export const SwitchOption: React.FC<SwitchOptionType> = ({
  title,
  className,
  onClick,
}) => {
  return (
    <li className={className} onClick={onClick}>
      {title}
    </li>
  );
};
