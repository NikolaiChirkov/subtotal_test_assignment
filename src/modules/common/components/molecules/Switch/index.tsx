import React from "react";
import styles from "./Switch.module.scss";
import { concatClassNames } from "@modules/common/utils";
import { SwitchOption } from "./SwitchOption";

type SwitchOptionType = {
  title: string;
  value: string;
  default: boolean;
};

export type SwitchType = {
  options: SwitchOptionType[];
  className?: string;
  onOptionChange: (value: string) => void;
};

export const Switch: React.FC<SwitchType> = ({ options, className, onOptionChange }) => {
  const [currentOption, setCurrentOption] = React.useState<string>(
    options.find((option) => option.default).title
  );

  return (
    <ul className={concatClassNames([styles.root, className])}>
      {options.map(({ title, value }) => (
        <SwitchOption
          key={title}
          title={title}
          className={concatClassNames([
            styles.root__option,
            currentOption === title && styles.current,
          ])}
          onClick={() => {
            onOptionChange(value);
            setCurrentOption(title);
          }}
        />
      ))}
    </ul>
  );
};
