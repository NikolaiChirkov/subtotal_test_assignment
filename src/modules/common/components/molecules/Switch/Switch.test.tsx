import React from "react";
import renderer from "react-test-renderer";
import { Switch } from "../Switch";

it("options changes classes when clicked", () => {
  const component = renderer.create(
    <Switch
      options={[
        { title: "Latest first", value: "asc", default: false },
        { title: "Newest first", value: "desc", default: true },
      ]}
      onOptionChange={(value) => console.log(value)}
    />
  );
  let tree = component.toJSON() as renderer.ReactTestRendererJSON;
  expect(tree).toMatchSnapshot();

  renderer.act(() => {
    tree.children[0].props.onClick();
  });

  tree = component.toJSON() as renderer.ReactTestRendererJSON;
  expect(tree).toMatchSnapshot();

  renderer.act(() => {
    tree.children[1].props.onClick();
  });
  
  tree = component.toJSON() as renderer.ReactTestRendererJSON;
  expect(tree).toMatchSnapshot();
});
