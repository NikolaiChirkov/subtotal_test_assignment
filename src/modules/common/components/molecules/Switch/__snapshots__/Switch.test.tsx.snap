// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options changes classes when clicked 1`] = `
<ul
  className=""
>
  <li
    className=""
    onClick={[Function]}
  >
    Latest first
  </li>
  <li
    className=""
    onClick={[Function]}
  >
    Newest first
  </li>
</ul>
`;

exports[`options changes classes when clicked 2`] = `
<ul
  className=""
>
  <li
    className=""
    onClick={[Function]}
  >
    Latest first
  </li>
  <li
    className=""
    onClick={[Function]}
  >
    Newest first
  </li>
</ul>
`;

exports[`options changes classes when clicked 3`] = `
<ul
  className=""
>
  <li
    className=""
    onClick={[Function]}
  >
    Latest first
  </li>
  <li
    className=""
    onClick={[Function]}
  >
    Newest first
  </li>
</ul>
`;
